livekit-agents/livekit/agents/pipeline/audio_recognition.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import asyncio
+from typing import TYPE_CHECKING, AsyncIterable, Literal, Protocol
+
+from livekit import rtc
+
+from .. import llm, stt, utils, vad
+from ..debug import tracing
+from ..log import logger
+from ..utils import aio
+from . import io
+
+if TYPE_CHECKING:
+    from .pipeline2 import PipelineAgent
+
+
+class _TurnDetector(Protocol):
+    # TODO: Move those two functions to EOU ctor (capabilities dataclass)
+    def unlikely_threshold(self) -> float: ...
+    def supports_language(self, language: str | None) -> bool: ...
+
+    async def predict_end_of_turn(self, chat_ctx: llm.ChatContext) -> float: ...
+
+
+EventTypes = Literal[
+    "start_of_speech",
+    "vad_inference_done",
+    "end_of_speech",
+    "interim_transcript",
+    "final_transcript",
+    "end_of_turn",
+]
+
+
+class AudioRecognition(rtc.EventEmitter[EventTypes]):
+    """
+    Audio recognition part of the PipelineAgent.
+    The class is always instantiated but no tasks may be running if STT/VAD is disabled
+
+    This class is also responsible for the end of turn detection.
+    """
+
+    UNLIKELY_END_OF_TURN_EXTRA_DELAY = 6.0
+
+    def __init__(
+        self,
+        *,
+        agent: PipelineAgent,
```

Comment:
would be better to take in a protocol that is a provider of chat context. would make it more easily testable.